### ISSUE TITLE
Reduce unneeded automatic debits/credits and problems past SHU date

### ISSUE THREAD

**Issue by kentbower** [kbower/tickettest1#2219]
_Imported from trac ticket 2219.
Created by **kbower**
Opened in trac: Mar 13, 2013, 7:27:23 AM
Last modified in trac: Jun 19, 2013, 9:24:31 AM_
***
This is all in attempt to reduce unneeded automatic debits/credits.
 * When applying discounts attempt to leave equivalent item discounts where they currently are.

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 14, 2013, 1:18:47 PM:_

This is also important to reduce need for #414.  See Morris ticket 0310333AJ45

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 14, 2013, 1:21:58 PM:_

Since the release of closing's been taken, this has come in the spot light and needs attention.

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 15, 2013, 9:52:29 AM:_

(Changeset [[3911]](../commit/4a23ba15d2bb0d40cb091885c1a82f27ef2d147a))
* Attempt to reduce unneeded automatic debits/credits and problems past SHU date (see #2219)
   * When applying discounts attempt to leave equivalent line discounts where they currently are.
   * sort consistently so that moving lines to other tickets is not likely to affect any discounted amount

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 19, 2013, 8:06:09 AM:_

(Changeset [[3925]](../commit/3153282dd95a3f0cd4e8720a92c7ed7a734cc28b))
* Add automatic adjusment comments to clarify why autoadjustment was created (see #2219)

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 19, 2013, 8:56:44 AM:_

In that specific example of 0315314KI15, because GERS sometimes uses NULL where Loft is explicit,
Loft saw these differences:

Closed line 6's "discount" was updated from [None] to [0]
Closed line 6's "takenwith" was updated from [None] to [N]

And interpreted these as meaningful differences.

I will work on catering to orders written in GERS so Loft understands there is no real difference there.

Thanks for pointing me there!
Kent

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 19, 2013, 8:58:05 AM:_

## Useful `paster shell` script
```python
def why(adjid):
    adj = DBSession.query(Order).get(adjid)
    order = adj.adjustedorder
    reasons = []
    def _why(adjustment, adjusted, accountingsign):
        ADJ_COLUMN_MAP = adjusted.__class__.ADJ_COLUMN_MAP
        AFFECTS_TOTAL = adjusted.__class__.AFFECTS_TOTAL
        for allow_concur_change, attr, name in adjusted.__class__.concurrent_check_cols:
            if allow_concur_change:
                continue
            if name in ADJ_COLUMN_MAP:
                adj_record_column = ADJ_COLUMN_MAP[name]
                if adj_record_column is None:
                    continue
            else:
                adj_record_column = name
            if name in AFFECTS_TOTAL:
                closed = getattr(adjusted, attr)
                modified = closed + getattr(adjustment, adj_record_column) * accountingsign
            else:
                closed = getattr(adjusted, attr)
                modified = getattr(adjustment, adj_record_column)
            if closed != modified:
                if isinstance(adjusted, OrderDetail):
                    reasons.append("""Closed line %s's "%s" was updated from [%s] to [%s]""" %
                        (adjusted.lineid, name, closed, modified))
                else:
                    reasons.append("""Closed "%s" was updated from [%s] to [%s]""" %
                        (name, closed, modified))
    _why(adj, order, adj.accountingsign)
    for l in adj.orderdetails:
        _why(l, obj_with_attr(order.orderdetails, 'lineid', int(l.autoadjusteduid)), adj.accountingsign)
    print "\n".join(reasons)

```

```
why(id)
```

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 19, 2013, 9:30:01 AM:_

```
03143A5QO16:
Closed line 4's "fillfromsiteid" was updated from [A5] to [None]
Closed line 4's "fillfromlocationid" was updated from [SHW] to [None]
Closed line 5's "discount" was updated from [None] to [0]

0313319QD44:
Closed line 7's "fillfromsiteid" was updated from [19] to [None]
Closed line 7's "fillfromlocationid" was updated from [SHW] to [None]
Closed line 8's "discount" was updated from [None] to [0]

03123A6PJ75:
Closed line 7's "discount" was updated from [None] to [0]


03073A3NG72:
Closed line 4's "discount" was updated from [None] to [0]

03013A3LE62:
Closed "recordgrouping" was updated from [None] to [AFS]
Closed "layawaysale" was updated from [None] to [N]
Closed "takewithsale" was updated from [None] to [N]
Closed line 1's "discount" was updated from [None] to [0]
Closed line 1's "deliverypoints" was updated from [None] to [0]
Closed line 2's "discount" was updated from [None] to [0]
Closed line 3's "discount" was updated from [None] to [0]
Closed line 5's "discount" was updated from [None] to [0]
Closed line 5's "deliverypoints" was updated from [None] to [0]

0301350LC21:
Closed "taxruleid" was updated from [401] to [440]
```
The majority at Ivan Smith are like the last...not sure what's up there. (see [#2235](2235))

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 20, 2013, 9:41:24 AM:_

(Changeset [[3927]](../commit/351c5dd70bec007d8a0bca5bb44c05c03cf66e14))
* eliminate most cases where $0 automatic adjustments (credits/debits) are created, though not needed (closes #2219)

**Comment by kentbower** [comment]
_**kbower** trac status change on Mar 20, 2013, 9:41:24 AM:_ **closed**
